File: serde-generate/runtime/java/serde/Tuple4.java
// Copyright (c) Facebook, Inc. and its affiliates
// SPDX-License-Identifier: MIT OR Apache-2.0

package serde;

public class Tuple4<T0, T1, T2, T3> {
    public T0 field0;
    public T1 field1;
    public T2 field2;
    public T3 field3;

    public Tuple4() {}

    public Tuple4(T0 f0, T1 f1, T2 f2, T3 f3) {
        this.field0 = f0;
        this.field1 = f1;
        this.field2 = f2;
        this.field3 = f3;
    }

    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null) return false;
        if (getClass() != obj.getClass()) return false;
        Tuple4 other = (Tuple4) obj;
        if (!this.field0.equals(other.field0)) { return false; }
        if (!this.field1.equals(other.field1)) { return false; }
        if (!this.field2.equals(other.field2)) { return false; }
        if (!this.field3.equals(other.field3)) { return false; }
        return true;
    }

    public int hashCode() {
        int value = 7;
        value = 31 * value + this.field0.hashCode();
        value = 31 * value + this.field1.hashCode();
        value = 31 * value + this.field2.hashCode();
        value = 31 * value + this.field3.hashCode();
        return value;
    }

}

File: serde-generate/runtime/dart/org/starcoin/serde/Tuple4.dart
import 'HashUtils.dart';

class Tuple4<T0, T1, T2, T3> {
  T0 field0;
  T1 field1;
  T2 field2;
  T3 field3;

  Tuple4(T0 f0, T1 f1, T2 f2, T3 f3) {
    assert(f0 != null);
    assert(f1 != null);
    assert(f2 != null);
    assert(f3 != null);
    this.field0 = f0;
    this.field1 = f1;
    this.field2 = f2;
    this.field3 = f3;
  }

  @override
  bool operator ==(covariant Tuple4 other) {
    if (this == other) return true;
    if (other == null) return false;
    if (this.field0 == other.field0 &&
        this.field1 == other.field1 &&
        this.field2 == other.field2 &&
        this.field3 == other.field3) {
      return true;
    } else {
      return false;
    }
  }

  @override
  int get hashCode => $jf($jc(
      $jc($jc(this.field0.hashCode, this.field1.hashCode),
          this.field2.hashCode),
      this.field3.hashCode));
}
